pkg/db/db_test.go

```diff
@@ -296,13 +308,11 @@ func TestClient_Download(t *testing.T) {
 			require.NoError(t, err, tc.name)
 
 			client := Client{
-				dbc:          mockConfig,
-				clock:        tc.clock,
 				githubClient: mockGitHubConfig,
```

Comment:
NIT: You can add arg to the constructor and do proper constructor dependency injection `db.NewClient(o GitHubOperation)`

Reply:
**This is just a discussion.**
 Actually, I'm always looking for the best way to inject dependencies.

This is a current implementation.

```
func NewClient() Client {
	return Client{
		dbc:          db.Config{},
		clock:        clock.RealClock{},
		githubClient: github.NewClient(),
	}
}
```

```
client := NewClient() // this is simple
```

If I expose the args, it is like the following.

```
func NewClient(config db.Config, clock clock.Clock, githubClient GitHubOperation) Client {
	return Client{
		dbc:           config,
		clock:        clock,
		githubClient: githubClient,
	}
}
```

And, it is called like this.
```
client := NewClient(db.Config{}, clock.RealClock{}, github.NewClient())
```

But these args are perfectly only for testing. A caller never change those args. In that case, I always wonder if they should be constructor arguments. I feel it is a bit too much.

If a user switches the value between MySQL, PostgreSQL, and Mock, it looks fine to me.

```
func NewClient(db *sql.DB) Client {
	return Client{
		db: db,
	}
}

// they are just examples.
client := NewClient(db.MySQL{})
client := NewClient(db.PostgreSQL{})
client := NewClient(MockDB{})
```

Reply:
I agree. We can keep it simple as is. Having constructor args and explicit dependency injection makes more sense when you develop a library that might be used in other places and leverage different strategies / interface implementations.

The only downside of referring directly to a private `githubClient` field rather than injecting it as a constructor args is exposing private field / implementation details in the test, which might be hard to maintain at some point. Imaging someone renaming private fields will cause the compilation failure of your tests. On the other hand, interfaces and constructor args won't break tests, i.e. you'd strictly test behaviour rather than implementation details.

Again the above consideration is general rule and sometime we can have exceptions to simplify things.

Reply:
I completely agree with you. That is a point. I believe we should test a behavior from outside rather than the implementation detail. If we use a private variable directly, it is not a test of behavior. From that point of view, they should be constructor arguments. But in that case, the constructor will have many arguments only for testing. It also increases the number of lines in a caller. There are pros and cons...
For now, I'll export the private variable as the argument.

Reply:
fixed